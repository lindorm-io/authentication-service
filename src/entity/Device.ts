import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { DeviceEvent } from "../enum";

export interface IDevicePIN {
  signature: string;
  updated: Date;
}

export interface IDevice extends IEntity {
  accountId: string;
  name: string;
  pin: IDevicePIN;
  publicKey: string;
  secret: string;
}

export interface IDeviceOptions extends IEntityBaseOptions {
  accountId: string;
  name?: string;
  pin?: IDevicePIN;
  publicKey: string;
  secret?: string;
}

export class Device extends EntityBase implements IDevice {
  private _accountId: string;
  private _name: string;
  private _pin: IDevicePIN;
  private _publicKey: string;
  private _secret: string;

  constructor(options: IDeviceOptions) {
    super(options);
    this._accountId = options.accountId;
    this._name = options.name || null;
    this._pin = {
      signature: options.pin?.signature || null,
      updated: options.pin?.updated || null,
    };
    this._publicKey = options.publicKey;
    this._secret = options.secret || null;
  }

  public get accountId(): string {
    return this._accountId;
  }

  public get name(): string {
    return this._name;
  }
  public set name(name: string) {
    this._name = name;
    this.addEvent(DeviceEvent.NAME_CHANGED, { name: this._name });
  }

  public get pin(): IDevicePIN {
    return this._pin;
  }
  public set pin(pin: IDevicePIN) {
    this._pin = pin;
    this.addEvent(DeviceEvent.PIN_CHANGED, { pin: this._pin });
  }

  public get publicKey(): string {
    return this._publicKey;
  }

  public get secret(): string {
    return this._secret;
  }
  public set secret(secret: string) {
    this._secret = secret;
    this.addEvent(DeviceEvent.SECRET_CHANGED, { secret: this._secret });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== DeviceEvent.CREATED) continue;
      throw new Error("Device has already been created");
    }
    this.addEvent(DeviceEvent.CREATED, {
      accountId: this._accountId,
      name: this._name,
      pin: this._pin,
      publicKey: this._publicKey,
      secret: this._secret,
      created: this._created,
      updated: this._updated,
    });
  }
}
