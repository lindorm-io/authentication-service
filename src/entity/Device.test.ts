import MockDate from "mockdate";
import { Device } from "./Device";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));

MockDate.set("2020-01-01 08:00:00.000");
const date = new Date("2020-01-01 08:00:00.000");

describe("Device.ts", () => {
  let device: Device;

  beforeEach(() => {
    device = new Device({
      accountId: "accountId",
      name: "name",
      pin: { signature: "pin", updated: date },
      publicKey: "publicKey",
      secret: "secret",
    });
  });

  test("should have all data", () => {
    expect(device).toMatchSnapshot();
  });

  test("should have optional data", () => {
    device = new Device({
      accountId: "accountId",
      publicKey: "publicKey",
    });

    expect(device).toMatchSnapshot();
  });

  test("should create", () => {
    device.create();
    expect(device.events).toMatchSnapshot();
  });

  test("should get accountId", () => {
    expect(device.accountId).toBe("accountId");
  });

  test("should get/set name", () => {
    expect(device.name).toBe("name");

    device.name = "new-name";

    expect(device.name).toBe("new-name");
    expect(device.events).toMatchSnapshot();
  });

  test("should get/set pin", () => {
    expect(device.pin).toStrictEqual({ signature: "pin", updated: date });

    device.pin = { signature: "new-pin", updated: date };

    expect(device.pin).toStrictEqual({ signature: "new-pin", updated: date });
    expect(device.events).toMatchSnapshot();
  });

  test("should get publicKey", () => {
    expect(device.publicKey).toBe("publicKey");
  });

  test("should get/set secret", () => {
    expect(device.secret).toBe("secret");

    device.secret = "new-secret";

    expect(device.secret).toBe("new-secret");
    expect(device.events).toMatchSnapshot();
  });
});
