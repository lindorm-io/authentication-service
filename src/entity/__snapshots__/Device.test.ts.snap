// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Device.ts should create 1`] = `
Array [
  Object {
    "created": 2020-01-01T07:00:00.000Z,
    "name": "device_created",
    "payload": Object {
      "accountId": "accountId",
      "created": 2020-01-01T07:00:00.000Z,
      "name": "name",
      "pin": Object {
        "signature": "pin",
        "updated": 2020-01-01T07:00:00.000Z,
      },
      "publicKey": "publicKey",
      "secret": "secret",
      "updated": 2020-01-01T07:00:00.000Z,
    },
  },
]
`;

exports[`Device.ts should get/set name 1`] = `
Array [
  Object {
    "created": 2020-01-01T07:00:00.000Z,
    "name": "device_name_changed",
    "payload": Object {
      "name": "new-name",
    },
  },
]
`;

exports[`Device.ts should get/set pin 1`] = `
Array [
  Object {
    "created": 2020-01-01T07:00:00.000Z,
    "name": "device_pin_changed",
    "payload": Object {
      "pin": Object {
        "signature": "new-pin",
        "updated": 2020-01-01T07:00:00.000Z,
      },
    },
  },
]
`;

exports[`Device.ts should get/set secret 1`] = `
Array [
  Object {
    "created": 2020-01-01T07:00:00.000Z,
    "name": "device_secret_changed",
    "payload": Object {
      "secret": "new-secret",
    },
  },
]
`;

exports[`Device.ts should have all data 1`] = `
Device {
  "_accountId": "accountId",
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [],
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_name": "name",
  "_pin": Object {
    "signature": "pin",
    "updated": 2020-01-01T07:00:00.000Z,
  },
  "_publicKey": "publicKey",
  "_secret": "secret",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;

exports[`Device.ts should have optional data 1`] = `
Device {
  "_accountId": "accountId",
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [],
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_name": null,
  "_pin": Object {
    "signature": null,
    "updated": null,
  },
  "_publicKey": "publicKey",
  "_secret": null,
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;
